packages/block-editor/src/utils/pre-parse-patterns.js

```diff
@@ -0,0 +1,53 @@
+/**
+ * WordPress dependencies
+ */
+import { useSelect, select } from '@wordpress/data';
+import { useEffect } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { store as blockEditorStore } from '../store';
+
+export function usePreParsePatterns() {
+	const patterns = useSelect(
+		( _select ) =>
+			_select( blockEditorStore ).getSettings()
+				.__experimentalBlockPatterns,
+		[]
+	);
+
+	useEffect( () => {
+		if ( ! patterns?.length ) {
+			return;
+		}
+
+		let handle;
+		let index = -1;
+
+		const callback = () => {
+			index++;
+			if ( index >= patterns.length ) {
+				return;
+			}
+
+			const marker = `[usePreParsePatterns] process ${ index } ${ patterns[ index ]?.title }`;
+			window.console.time( marker );
+			select( blockEditorStore ).__experimentalGetParsedBlocks(
+				patterns[ index ].content,
+				index
+			);
+			window.console.timeEnd( marker );
+
+			handle = window.requestIdleCallback( callback );
```

Comment:
we have the priority query package that seems perfect for these idle callback queues. Did you consider it?

Reply:
Yes, I tried it. One problem I ran into is priority queue can't limit how many tasks to run per callback. It always tries to use up all the remaining processing time which didn't work quite well. It was noticeably laggier when opening up the editor because it tried to parse too many patterns. Since `parse` might take more than the remaining idle time.